src/backend/distributed/utils/causal_clock.c

```diff
@@ -0,0 +1,787 @@
+		}
+
+		uint64 logical = ParseIntField(result, 0, 0);
+		uint32 counter = ParseIntField(result, 0, 1);
+		uint64 nodeClockValue = SET_CLOCK(logical, counter);
+
+		ereport(DEBUG1, (errmsg(
+							 "Node(%lu:%u) transaction clock %lu:%u",
+							 connection->connectionId, connection->port,
+							 logical, counter)));
+
+		if (nodeClockValue > globalClockValue)
+		{
+			globalClockValue = nodeClockValue;
+		}
+
+		PQclear(result);
+		ForgetResults(connection);
+	}
+
+	ereport(DEBUG1,
+			(errmsg("Final global transaction clock %lu:%u",
+					GET_LOGICAL(globalClockValue),
+					(uint32) GET_COUNTER(globalClockValue))));
+
+	return globalClockValue;
+}
+
+
+/*
+ * AdjustClocksToTransactionHighest Sets the clock value of all the nodes, participated
+ * in the PREPARE of the transaction, to the highest clock value of all the nodes.
+ */
+static void
+AdjustClocksToTransactionHighest(List *nodeConnectionList, uint64 transactionClockValue)
+{
+	StringInfo queryToSend = makeStringInfo();
+	uint64 transactionLogicalClock = GET_LOGICAL(transactionClockValue);
+	uint32 transactionCounterClock = GET_COUNTER(transactionClockValue);
+
+	/* Set the adjusted value locally */
+	AdjustLocalClockToRemote(transactionLogicalClock, transactionCounterClock);
+
+	/* Set the clock value on participating worker nodes */
+	MultiConnection *connection = NULL;
+	resetStringInfo(queryToSend);
+	appendStringInfo(queryToSend,
+					 "SELECT "
+					 "pg_catalog.citus_internal_adjust_local_clock_to_remote(ROW(%lu, %u));",
+					 transactionLogicalClock, transactionCounterClock);
+
+	foreach_ptr(connection, nodeConnectionList)
+	{
+		int querySent = SendRemoteCommand(connection, queryToSend->data);
+
+		if (querySent == 0)
+		{
+			ReportConnectionError(connection, ERROR);
+		}
+	}
+
+	/* Process the result */
+	foreach_ptr(connection, nodeConnectionList)
+	{
+		bool raiseInterrupts = true;
+		PGresult *result = GetRemoteCommandResult(connection, raiseInterrupts);
+
+		if (!IsResponseOK(result))
+		{
+			ReportResultError(connection, result, ERROR);
+		}
+
+		PQclear(result);
+		ForgetResults(connection);
+	}
+}
+
+
+/*
+ * During prepare, once all the nodes acknowledge commit, persist the current
+ * transaction id along with the clock value in the catalog.
+ */
+void
+PrepareAndSetTransactionClock(List *transactionNodeList)
+{
+	if (!EnableClusterClock)
+	{
+		/* citus.enable_cluster_clock is false */
+		return;
+	}
+
+	/*
+	 * Call get_current_transaction_id UDF to get the current
+	 * distributed transaction id.
+	 */
+	Oid transactionFuncOid = FunctionOid("pg_catalog", "get_current_transaction_id", 0);
+	Datum transactionIdHeapDatum = OidFunctionCall0(transactionFuncOid);
+
+	FmgrInfo *outputFunction = (FmgrInfo *) palloc0(sizeof(FmgrInfo));
+	Oid outputFunctionId = FunctionOid("pg_catalog", "record_out", 1);
+	fmgr_info(outputFunctionId, outputFunction);
+	char *transactionId = OutputFunctionCall(outputFunction, transactionIdHeapDatum);
```

Comment:
Would it make sense to add a function in C that does the exact formatting of the local transaction id here. That is asssuming we don't already have one.

It seems that the contents of the xact id is all available from `GetCurrentDistributedTransactionId();`. It is quite hard to understand what we are doing here by calling two udf's based on their Oid. Is this how we generally get a text representation for the xact id?

if that is the case, lets atleast extract this into something documented and reusable.

Reply:
If we have a generic routine to get text representation of xact id, that will be the best option, but I did all sorts of code search and couldn't find one.  As we already get_current_transaction_id() and used in our code 
`#define GET_PROCESS_ID "SELECT process_id FROM get_current_transaction_id()"`

I will make this as a reusable one.

Reply:
We recommend applications to save the changes along with the xact id `get_current_transaction_id()`, so that it can join with the `pg_dist_commit_transaction` catalog to get the commit clock. I've put the code in a generic routine.